src/include/miopen/pad_constant/problem_description.hpp

```
@@ -0,0 +1,209 @@
+                       const size_t* padding_,
+                       const int padding_size_ = 0)
+        : xDesc(xDesc_), yDesc(yDesc_), padding(padding_), padding_size(padding_size_)
+    {
+        if(!IsPaddingValid())
+            MIOPEN_THROW("Padding is not valid");
+        if(!IsSameShape())
+            MIOPEN_THROW("Tensors do not have the same shapes");
+        if(!IsSameType())
+            MIOPEN_THROW("Tensor values do not have the same type");
+    }
+
+    const TensorDescriptor& GetXDesc() const { return xDesc; }
+    const TensorDescriptor& GetYDesc() const { return yDesc; }
+
+    NetworkConfig MakeNetworkConfig() const override;
+
+    bool IsSameType() const
+    {
+        if(xDesc.GetType() == yDesc.GetType())
+            return true;
+        return false;
+    }
+
+    bool IsSameShape() const
+    {
+        auto xSize = xDesc.GetSize();
+        auto ySize = yDesc.GetSize();
+        if(xSize == ySize)
+            return true;
+        return false;
+    }
+
+    bool IsContiguous() const { return xDesc.IsContiguous() && yDesc.IsContiguous(); }
+
+    bool IsImprovementOverROCm() const
+    {
+        if(IsContiguous())
+        {
+            // Contiguous case
+            // No winning over ROCm, only ~7% of tested cases get > 20% improvement
+            return false;
+        }
+        else
+        {
+            // Non contiguous case
+            // We win over ROCm unless the only padded dimension is n
+            bool all_zero = true;
+            for(int i = 0; i < 10; i++)
+            {
+                if(i == 9 || i == 8)
+                {
+                    if(all_zero && padding[i] != 0)
+                        return false;
+                }
+                else
+                {
+                    if(padding[i] != 0)
+                        all_zero = false;
+                }
+            }
+            return true;
+        }
+    }
+
+    bool IsPaddingValid() const
```

If the input value is not valid, throw rather than return false.